core/services/offchainreporting2relay/delegate.go

```diff
@@ -0,0 +1,275 @@
+package offchainreporting2relay
+
+import (
+	"math/big"
+	"time"
+
+	"github.com/smartcontractkit/sqlx"
+
+	"github.com/smartcontractkit/libocr/offchainreporting2/reportingplugin/median/evmreportcodec"
+
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/bridges"
+	"github.com/smartcontractkit/chainlink/core/chains/evm"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/keystore"
+	"github.com/smartcontractkit/chainlink/core/services/keystore/keys/p2pkey"
+	"github.com/smartcontractkit/chainlink/core/services/ocrcommon"
+	"github.com/smartcontractkit/chainlink/core/services/offchainreporting2"
+	"github.com/smartcontractkit/chainlink/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/core/services/relay"
+	"github.com/smartcontractkit/chainlink/core/services/relay/solana"
+	"github.com/smartcontractkit/chainlink/core/services/telemetry"
+	ocr "github.com/smartcontractkit/libocr/offchainreporting2"
+	"github.com/smartcontractkit/libocr/offchainreporting2/reportingplugin/median"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2/types"
+)
+
+type Delegate struct {
+	db                    *sqlx.DB
+	jobORM                job.ORM
+	keyStore              keystore.OCR2
+	pipelineRunner        pipeline.Runner
+	peerWrapper           *ocrcommon.SingletonPeerWrapper
+	monitoringEndpointGen telemetry.MonitoringEndpointGenerator
+	chainSet              evm.ChainSet
+	lggr                  logger.Logger
+	relays                relay.Relays
+}
+
+var _ job.Delegate = (*Delegate)(nil)
+
+func NewDelegate(
+	db *sqlx.DB,
+	jobORM job.ORM,
+	keyStore keystore.OCR2,
+	pipelineRunner pipeline.Runner,
+	peerWrapper *ocrcommon.SingletonPeerWrapper,
+	monitoringEndpointGen telemetry.MonitoringEndpointGenerator,
+	chainSet evm.ChainSet,
+	lggr logger.Logger,
+	relays relay.Relays,
+) *Delegate {
+	return &Delegate{
+		db,
+		jobORM,
+		keyStore,
+		pipelineRunner,
+		peerWrapper,
+		monitoringEndpointGen,
+		chainSet,
+		lggr,
+		relays,
+	}
+}
+
+func (d Delegate) JobType() job.Type {
+	return job.OffchainReporting2Relay
+}
+
+func (Delegate) OnJobCreated(spec job.Job) {}
+func (Delegate) OnJobDeleted(spec job.Job) {}
+
+func (Delegate) AfterJobCreated(spec job.Job)  {}
+func (Delegate) BeforeJobDeleted(spec job.Job) {}
+
+func (d Delegate) ServicesForSpec(jobSpec job.Job) (services []job.Service, err error) {
+	if jobSpec.Offchainreporting2OracleSpec == nil {
+		return nil, errors.Errorf("offchainreporting.Delegate expects an *job.Offchainreporting2OracleSpec to be present, got %v", jobSpec)
+	}
+	spec := jobSpec.Offchainreporting2OracleSpec
+
+	chain, err := d.chainSet.Get(jobSpec.Offchainreporting2OracleSpec.EVMChainID.ToInt())
+	if err != nil {
+		return nil, err
+	}
+
+	ocrdb := offchainreporting2.NewDB(d.db.DB, spec.ID)
+
+	// TODO [relay]: make a relay choice depending on job spec
```

Comment:
Might be cleaner to make this choice inside the relay service itself? I think we can make the ocr2-relay job type a generic relay job type, that passes a relay type field and a json field of relay-config (containing the additional account metadata for solana and who knows what else for other chains) straight through to a generic relay service. Basically what you have already except NewOCR2Service just accepts a spec.relayType and spec.relayConfig and we do the switch inside an object that implements Relay